pkg/cmd/membership/membership.go

```diff
@@ -0,0 +1,168 @@
+func NewMembershipCommand(errOut io.Writer) *cobra.Command {
+	membership := &membership{
+		errOut: errOut,
+	}
+	cmd := &cobra.Command{
+		Use:   "membership",
+		Short: "cluster membership validate",
+		Long:  "This command makes sure that the local etcd member is part of the cluster.",
+		Run: func(cmd *cobra.Command, args []string) {
+			must := func(fn func() error) {
+				if err := fn(); err != nil {
+					if cmd.HasParent() {
+						klog.Fatal(err)
+						fmt.Fprint(membership.errOut, err.Error())
+					}
+				}
+			}
+			must(membership.Complete)
+			must(membership.Validate)
+			must(membership.Run)
+		},
+	}
+	membership.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+func (m *membership) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&m.EtcdDnsName, "etcd-dns-name", "", "etcd member FQDN")
+	fs.StringVar(&m.EtcdEndpoints, "endpoints", "", "commas seperated list of etcd clients")
+	fs.StringVar(&m.etcdDataDir, "data-dir", "/var/lib/etcd", "path to etcd data-dir")
+}
+func (m *membership) Validate() error {
+	if m.IsExistingMember() {
+		if m.EtcdEndpoints == "" || m.EtcdDnsName == "" {
+			return fmt.Errorf("membership validation requires etcd-dns-name and endpoints")
+		}
+	}
+	return nil
+}
+
+func (m *membership) Complete() error {
+	m.EtcdDataFile = filepath.Join(m.etcdDataDir, dbPath)
+
+	etcdDnsName, etcdEndpoints := getENV()
+	// flags take priority over ENV if used.
+	if m.EtcdDnsName != "" && etcdDnsName != "" {
+		klog.Infof("flag --etcd-dns-name and ENV ETCD_DNS_NAME set, ENV will be ignored")
+	}
+	if m.EtcdEndpoints != "" && etcdEndpoints != "" {
+		klog.Infof("flag --endpoints and ENV ETCDCTL_ENDPOINTS set, ENV will be ignored")
+	}
+
+	if m.EtcdEndpoints == "" && etcdEndpoints != "" {
+		klog.Infof("recognized and used environment variable ETCDCTL_ENDPOINTS=%s", etcdEndpoints)
+		m.EtcdEndpoints = etcdEndpoints
+	}
+	if m.EtcdDnsName == "" && etcdDnsName != "" {
+		klog.Infof("recognized and used environment variable ETCD_DNS_NAME=%s", etcdDnsName)
+		m.EtcdDnsName = etcdDnsName
+	}
+	return nil
+}
+
+func (m *membership) Run() error {
+	if m.IsExistingMember() {
+		klog.Infof("etcd has previously been initialized as a member")
+		return nil
+	}
+
+	cli, err := getEtcdClient(strings.Split(m.EtcdEndpoints, ","))
+	if err != nil {
+		return err
+	}
+	defer cli.Close()
+
+	wait.PollInfinite(retryDuration, func() (bool, error) {
+		ctx, cancel := context.WithCancel(context.Background())
+		l, err := cli.MemberList(ctx)
+		cancel()
+		if err != nil {
+			klog.Errorf("memberlist failed: %v", err)
+			return false, nil
+		}
+
+		//TODO we should handle IP based peers as well
+		for _, member := range l.Members {
+			if member.Name == m.EtcdDnsName {
+				klog.Infof("%s found in cluster", m.EtcdDnsName)
+				return true, nil
+			}
+		}
+		klog.Infof("%s is currently not a member of the cluster %+v", m.EtcdDnsName, l.Members)
+		return false, nil
+	})
+
+	return nil
+}
+func (m *membership) IsExistingMember() bool {
```

Comment:
why have this function, it simply calls the underlying function, can we use `waitforkube.CheckEtcdDataFileExists(m.EtcdDataFile)` instead?

Reply:
and probably import waitforkube pack as `util` 📦 so it reads better

Reply:
just readability is all, yes good idea on util